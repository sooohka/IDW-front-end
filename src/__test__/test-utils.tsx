/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable no-param-reassign */
import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import renderer from "react-test-renderer";
import store from "../store";
import GlobalStyle from "../style/GlobalStyle";
import theme from "../style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const thunk =
  ({ dispatch, getState }: any) =>
  (next: any) =>
  (action: any) => {
    if (typeof action === "function")
      // 여기서 반환되는 action은 thunk함수임
      return action(dispatch, getState);
    return next(action);
  };

const useStore = (stateName: string, initState: any, reducer: any) => {
  const store = {
    dispatch: jest.fn((action: any) => {
      initState = reducer(initState, action);
    }),
    getState: jest.fn(() => ({ [stateName]: initState })),
  };
  const next = jest.fn();
  const makeThunk = (action: any) => thunk(store)(next)(action);
  return { store, next, makeThunk };
};

const createWithThemeProvider = (EL: ReactElement) => {
  return renderer.create(
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {EL}
    </ThemeProvider>,
  );
};

const redux = { thunk, useStore };
export { redux, customRender as render, createWithThemeProvider };

export * from "@testing-library/react";
