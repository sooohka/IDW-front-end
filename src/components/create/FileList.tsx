import React, { useEffect, useRef, useState } from "react";
import styled, { css } from "styled-components";
import { ReactComponent as Sort } from "../../assets/icons/sort-up-solid.svg";

const StyledFileList = styled.div`
  margin: 1rem 0 0 0;
  border: 2px solid black;
  border-radius: 5px;
  min-height: 3rem;
  position: relative;
`;

const FileListTitle = styled.h3`
  flex: 1;
  padding-left: 1rem;
`;

const FileListHeader = styled.div`
  display: flex;
  align-items: center;
  cursor: pointer;
`;

const Files = styled.div<{ isFolded: boolean }>`
  transition: all;
  transform-origin: top;
  ${({ isFolded }) => {
    if (isFolded)
      return css`
        transform: scaleY(0);
        max-height: 0;
      `;
    return css`
      transform: scaleY(1);
      padding: 2rem 0 0;
      max-height: 10000px;
    `;
  }}
`;

const IconWrapper = styled.div<{ isFolded: boolean }>`
  cursor: pointer;
  & > svg {
    transition: all 0.2s;
    ${({ isFolded }) => {
      if (isFolded) {
        return css`
          transform: rotate(180deg);
        `;
      }
      return css`
        transform: rotate(0deg) translateY(30%);
      `;
    }}
  }
`;

interface IProps {
  children: React.ReactNode;
}

const FileList: React.FC<IProps> = ({ children }) => {
  const [isFolded, setIsFolded] = useState(false);
  const bottomEl = useRef<HTMLDivElement>(null);

  const handleFoldBtnClick = () => {
    setIsFolded((prev) => !prev);
  };
  useEffect(() => {
    if (isFolded) window.scrollTo({ behavior: "smooth", top: 0 });
    else bottomEl.current?.scrollIntoView({ behavior: "smooth", block: "center" });
  }, [isFolded]);
  return (
    <StyledFileList>
      <FileListHeader onClick={handleFoldBtnClick}>
        <FileListTitle>파일들</FileListTitle>
        <IconWrapper isFolded={isFolded}>
          <Sort width={30} height={30} />
        </IconWrapper>
      </FileListHeader>
      <Files isFolded={isFolded}>{children}</Files>
      <div ref={bottomEl} style={{ display: "flex", justifyContent: "flex-end" }}>
        <IconWrapper onClick={handleFoldBtnClick} isFolded={isFolded}>
          go Top
          <Sort width={30} height={30} />
        </IconWrapper>
      </div>
    </StyledFileList>
  );
};

export default FileList;
