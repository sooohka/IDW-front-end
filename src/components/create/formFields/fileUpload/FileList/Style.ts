import styled, { css } from "styled-components";

const FileList = styled.div`
  margin: 1rem 0 0 0;
  border: 2px solid black;
  border-radius: 5px;
  min-height: 3rem;
  position: relative;
`;

const Title = styled.h3`
  flex: 1;
  padding-left: 1rem;
`;

const Header = styled.div`
  display: flex;
  align-items: center;
  cursor: pointer;
`;

const Files = styled.div<{ isFolded: boolean }>`
  transition: all;
  transform-origin: top;
  ${({ isFolded }) => {
    if (isFolded)
      return css`
        transform: scaleY(0);
        max-height: 0;
      `;
    return css`
      transform: scaleY(1);
      padding: 2rem 0 0;
      max-height: 10000px;
    `;
  }}
`;

const IconWrapper = styled.div<{ isFolded: boolean }>`
  cursor: pointer;
  & > svg {
    transition: all 0.2s;
    ${({ isFolded }) => {
      if (isFolded) {
        return css`
          transform: rotate(180deg);
        `;
      }
      return css`
        transform: rotate(0deg) translateY(30%);
      `;
    }}
  }
`;

const NotFoundText = styled.p`
  text-align: center;
`;

export { FileList, Header, Title, NotFoundText, Files, IconWrapper };
