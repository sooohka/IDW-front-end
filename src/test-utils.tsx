import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import store from "./store";
import GlobalStyle from "./style/GlobalStyle";
import theme from "./style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
