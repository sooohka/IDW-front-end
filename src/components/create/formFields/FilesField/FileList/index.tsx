import React, { useLayoutEffect, useRef, useState } from "react";
import { ReactComponent as Sort } from "../../../../../assets/icons/sort-up-solid.svg";
import Text from "../../../../common/Text";
import * as S from "./Style";

const FileList: React.FC = ({ children }) => {
  const [isFolded, setIsFolded] = useState(true);
  const bottomEl = useRef<HTMLDivElement>(null);

  const handleFoldBtnClick = () => {
    setIsFolded((prev) => !prev);
  };

  useLayoutEffect(() => {
    if (isFolded) bottomEl.current?.scrollIntoView({ behavior: "smooth", block: "end" });
    else bottomEl.current?.scrollIntoView({ behavior: "smooth", block: "center" });
  }, [isFolded]);

  return (
    <S.FileList>
      <S.Header onClick={handleFoldBtnClick}>
        <S.Title>파일들</S.Title>
        <S.IconWrapper isFolded={isFolded}>
          <Sort width={30} height={30} />
        </S.IconWrapper>
      </S.Header>
      <S.Files isFolded={isFolded}>
        {React.Children.count(children) === 0 ? (
          <div style={{ textAlign: "center" }}>
            <Text bold fontSize='strongBody' text='파일이 없습니다' />
          </div>
        ) : (
          children
        )}
      </S.Files>
      <div ref={bottomEl} style={{ display: "flex", justifyContent: "flex-end" }}>
        <button type='button' onClick={handleFoldBtnClick}>
          Fold
        </button>
      </div>
    </S.FileList>
  );
};

export default FileList;
