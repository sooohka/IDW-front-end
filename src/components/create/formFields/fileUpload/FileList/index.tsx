import React, { useLayoutEffect, useRef, useState } from "react";
import { ReactComponent as Sort } from "../../../../../assets/icons/sort-up-solid.svg";
import useMount from "../../../../../utils/hooks/useMount";
import Text from "../../../../common/Text";
import FileUploadWithProgress from "../FileUploadWIthProgress";
import * as S from "./Style";

interface IProps {
  imageFiles: TargetFile[];
  handleDelete: (id: string) => () => void;
  setImageFiles: React.Dispatch<React.SetStateAction<TargetFile[]>>;
}

const FileList: React.FC<IProps> = ({ imageFiles, handleDelete, setImageFiles }) => {
  const [isFolded, setIsFolded] = useState(true);
  const { isMount } = useMount();
  const bottomEl = useRef<HTMLDivElement>(null);

  const handleFoldBtnClick = () => {
    setIsFolded((prev) => !prev);
  };

  useLayoutEffect(() => {
    if (isMount) return;
    if (isFolded) bottomEl.current?.scrollIntoView({ behavior: "smooth", block: "end" });
    else bottomEl.current?.scrollIntoView({ behavior: "smooth", block: "center" });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isFolded]);

  useLayoutEffect(() => {
    if (imageFiles.length > 0) setIsFolded(false);
  }, [imageFiles]);

  return (
    <S.FileList>
      <S.Header onClick={handleFoldBtnClick}>
        <S.Title>파일들</S.Title>
        <S.IconWrapper isFolded={isFolded}>
          <Sort width={30} height={30} />
        </S.IconWrapper>
      </S.Header>
      <S.Files isFolded={isFolded}>
        {imageFiles.length !== 0 ? (
          imageFiles.map((imageFile) => (
            <FileUploadWithProgress
              imageFile={imageFile}
              key={imageFile.id}
              handleDelete={handleDelete}
              setImageFiles={setImageFiles}
            />
          ))
        ) : (
          <div style={{ textAlign: "center" }}>
            <Text bold fontSize='strongBody' text='파일이 없습니다' />
          </div>
        )}
      </S.Files>
      <div ref={bottomEl} style={{ display: "flex", justifyContent: "flex-end" }}>
        {imageFiles.length > 5 ? (
          <S.IconWrapper onClick={handleFoldBtnClick} isFolded={isFolded}>
            go Top
            <Sort width={30} height={30} />
          </S.IconWrapper>
        ) : null}
      </div>
    </S.FileList>
  );
};

export default FileList;
