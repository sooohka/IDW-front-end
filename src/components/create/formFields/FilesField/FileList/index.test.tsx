import userEvent from "@testing-library/user-event";
import FileList from ".";
import { act, screen, render, createWithThemeProvider } from "../../../../../test/test-utils";

describe("src/components/create/formFields/FilesField/FileList", () => {
  it("renders", () => {
    const component = createWithThemeProvider(<FileList />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("render '파일이 없습니다' when has no children", async () => {
    render(<FileList />);

    const textEl = screen.getByText(/파일이 없습니다/i);
    expect(textEl).toBeInTheDocument();
  });

  it("doesn't render '파일이 없습니다 when has children", () => {
    const children = [1, 2, 3, 4, 5];
    render(
      <FileList>
        {children.map((v) => (
          <div key={v}>{v}</div>
        ))}
      </FileList>,
    );

    const textEl = screen.queryByText(/파일이 없습니다/i);
    expect(textEl).not.toBeInTheDocument();
  });

  it("fold when fold button clicked", () => {
    const files = [1, 2, 3];
    render(
      <FileList>
        {files.map((v) => (
          <div key={v}>{v}</div>
        ))}
      </FileList>,
    );
    const fileListHeaderEl = screen.getByTestId(/file-list-header/i);
    const fileListContent = screen.queryByTestId(/file-list-content/i);
    const textEl = screen.queryByText(/파일이 없습니다/i);

    userEvent.click(fileListHeaderEl);

    expect(fileListContent).not.toBeInTheDocument();
    expect(textEl).not.toBeInTheDocument();
  });

  it("unfold when fold button clicked", () => {
    const files = [1, 2, 3];
    render(
      <FileList>
        {files.map((v) => (
          <div key={v}>{v}</div>
        ))}
      </FileList>,
    );
    const fileListHeaderEl = screen.getByTestId(/file-list-header/i);
    const fileListContent = screen.queryByTestId(/file-list-content/i);

    act(() => {
      userEvent.click(fileListHeaderEl);
      userEvent.click(fileListHeaderEl);
    });

    expect(fileListContent).toBeInTheDocument();
  });
});
