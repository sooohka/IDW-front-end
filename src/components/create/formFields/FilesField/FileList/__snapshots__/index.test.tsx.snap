// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/create/formFields/FilesField/FileList renders 1`] = `
<div
  className="sc-gsDKAQ iZkqjs"
>
  <div
    className="sc-hKwDye hmZybT"
    data-testid="file-list-header"
    onClick={[Function]}
  >
    <h3
      className="sc-dkPtRN dsURgc"
    >
      파일들
    </h3>
    <div
      className="sc-jRQBWg ermYcA"
    >
      <svg
        height={30}
        width={30}
      >
        sort-up-solid.svg
      </svg>
    </div>
  </div>
  <div
    className="sc-eCImPb cMmyzq"
    data-testid="file-list-content"
  >
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <p
        className="sc-bdvvtL hEkgRH"
        fontSize="strongBody"
      >
        파일이 없습니다
      </p>
    </div>
  </div>
</div>
`;
